retriever/retriever.go

```diff
@@ -0,0 +1,116 @@
+// Package retriever wraps functionality for fetching data from filecoin.
+package retriever
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/data-preservation-programs/singularity/retriever/deserializer"
+	lassietypes "github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/storage"
+	trustlessutils "github.com/ipld/go-trustless-utils"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multicodec"
+	"go.uber.org/multierr"
+)
+
+// EndpointFinder finds http endpoints for set of SPs
+type EndpointFinder interface {
+	FindHTTPEndpoints(ctx context.Context, sps []string) ([]peer.AddrInfo, error)
+}
+
+// Retriever wraps Lassie and HTTP endpoint lookup to returns a byte stream for
+// a cid at the root of a unixfs tree, from a list of Filecoin providers
+type Retriever struct {
+	lassie         lassietypes.Fetcher
+	endpointFinder EndpointFinder
+}
+
+// NewRetriever returns a new retriever instance
+func NewRetriever(lassie lassietypes.Fetcher, endpointFinder EndpointFinder) *Retriever {
+	return &Retriever{
+		lassie:         lassie,
+		endpointFinder: endpointFinder,
+	}
+}
+
+// deserialize takes an reader of a carFile and writes the deserialized output
+func (r *Retriever) deserialize(ctx context.Context, c cid.Cid, rangeStart int64, rangeEnd int64, carInput io.Reader, carOutput io.Writer) error {
+	cr, err := car.NewBlockReader(carInput)
+	if err != nil {
+		return err
+	}
+	_, err = deserializer.Deserialize(ctx, cr, c, rangeStart, rangeEnd, carOutput)
+	return err
+}
+
+// getContent fetches content through Lassie and writes the CAR file to an output writer
+func (r *Retriever) getContent(ctx context.Context, c cid.Cid, rangeStart int64, rangeEnd int64, sps []string, carOutput io.Writer) error {
+	writable, err := storage.NewWritable(carOutput, []cid.Cid{c}, car.WriteAsCarV1(true))
+	if err != nil {
+		return err
+	}
+	providerAddrs, err := r.endpointFinder.FindHTTPEndpoints(ctx, sps)
+	if err != nil {
+		return err
+	}
+	inclusiveRangeEnd := rangeEnd - 1
+	request, err := lassietypes.NewRequestForPath(writable, c, "", trustlessutils.DagScopeEntity, &trustlessutils.ByteRange{
+		From: rangeStart,
+		// byte range is INCLUSIVE in the lassie/trustless HTTP context, so decrement
+		To: &inclusiveRangeEnd,
+	})
+	if err != nil {
+		return err
+	}
+	request.Duplicates = true
+	request.Protocols = []multicodec.Code{multicodec.TransportIpfsGatewayHttp}
```

Comment:
Have you evaluated usinig piece gateway, because each fileRange is stored sequentially in the file, so you can get the CarBlocks of a fileRange, and stream that fileRange with a single HTTP request with ranges header while validating the content while streaming.

Reply:
Or, we should make it configurable so the user can choose different ways of retrieval based on some header value, i.e. x-trasnport-protocol